pandas/core/groupby.py

```diff
@@ -3740,9 +3740,31 @@ def _reindex_output(self, result):
             return result
 
         levels_list = [ping.group_index for ping in groupings]
-        index = MultiIndex.from_product(levels_list, names=self.grouper.names)
-        d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
-        return result.reindex(**d).sortlevel(axis=self.axis)
+        index, _ = MultiIndex.from_product(
+            levels_list, names=self.grouper.names).sortlevel()
+
+        if self.as_index:
+            d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
+            return result.reindex(**d)
+
+        # GH 13204
```

Comment:
In this case, this is correct (it's slightly slower because of dropping and inserting columns to the data frame - but overall much cleaner).

In a general case, not all groupers have to be columns. So, probably, we have to manually find which groupers are columns and remove them from `result`, which will make the code even more obscure.

Something like this
```
result.set_index(self.grouper.result_index).reindex(index).reset_index()
```
fails because `reset_index` raises `ValueError: cannot insert A, already exists`. 

And there is also a (probably rare but not unlikely) corner case, where a non-column grouper has the same name as some other (non-grouping) column in `result`.